packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts

```diff
@@ -0,0 +1,244 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { getDataChunker } from './getDataChunker';
+import { S3UploadOptions } from '../../../types';
+import { resolveS3ConfigAndInput } from '../../../utils';
+import {
+	abortMultipartUpload,
+	completeMultipartUpload,
+	headObject,
+	Part,
+} from '../../../../../AwsClients/S3';
+import { StorageUploadDataRequest } from '../../../../../types';
+import { S3Item } from '../../../types/results';
+import {
+	DEFAULT_ACCESS_LEVEL,
+	DEFAULT_QUEUE_SIZE,
+} from '../../../utils/constants';
+import { loadOrCreateMultipartUpload } from './initialUpload';
+import { ResolvedS3Config } from '../../../types/options';
+import { getConcurrentUploadsProgressTracker } from './progressTracker';
+import { getUploadsCacheKey, removeCachedUpload } from './uploadCache';
+import { uploadPartExecutor } from './uploadPartExecutor';
+import { StorageError } from '../../../../../errors/StorageError';
+import { AmplifyV6, StorageAccessLevel } from '@aws-amplify/core';
+
+/**
+ * Create closure hiding the multipart upload implementation details and expose the upload job and control functions(
+ * onPause, onResume, onCancel).
+ *
+ * @internal
+ */
+export const getMultipartUploadHandlers = (
+	{
+		options: uploadDataOptions,
+		key,
+		data,
+	}: StorageUploadDataRequest<S3UploadOptions>,
+	totalLength?: number
+) => {
+	let resolveCallback: ((value: S3Item) => void) | undefined;
+	let rejectCallback: ((reason?: any) => void) | undefined;
+	let inProgressUpload:
+		| {
+				uploadId: string;
+				completedParts: Part[];
+		  }
+		| undefined;
+	let s3Config: ResolvedS3Config | undefined;
+	let abortController: AbortController | undefined;
+	let bucket: string | undefined;
+	let keyPrefix: string | undefined;
+
+	const startUpload = async (): Promise<S3Item> => {
+		const resolvedS3Options = await resolveS3ConfigAndInput(uploadDataOptions);
+		s3Config = resolvedS3Options.s3Config;
+		bucket = resolvedS3Options.bucket;
+		keyPrefix = resolvedS3Options.keyPrefix;
```

Comment:
nit: can we do object destructuring here?
```suggestion
		const { s3Config, bucket , keyPrefix } = await resolveS3ConfigAndInput(uploadDataOptions);
```

Reply:
Actually no, the `s3Config`, `bucket`, `keyPrefix` variables are initiated in the closure. Here I'm trying to set these variables so other functions in the same closure can access these variables.

Reply:
Got it thanks!!